tests/openassetio/test_traits.py

```diff
@@ -0,0 +1,158 @@
+        assert isinstance(a_specification.getTraitProperty("second_trait", "an int"), int)
+        assert a_specification.getTraitProperty("first_trait", "a float") == 1.0
+        assert isinstance(a_specification.getTraitProperty("first_trait", "a float"), float)
+        assert a_specification.getTraitProperty("second_trait", "a bool") is True
+
+    def test_when_key_is_not_found_then_get_returns_None(self, a_specification):
+        assert a_specification.getTraitProperty("first_trait", "a string") is None
+
+    def test_when_trait_is_not_found_then_IndexError_raised(self, a_specification):
+        with pytest.raises(IndexError):
+            a_specification.setTraitProperty("unknown_trait", "a string", "string")
+
+        with pytest.raises(IndexError):
+            _ = a_specification.getTraitProperty("unknown_trait", "a string")
+
+    def test_when_value_is_not_supported_then_set_raises_TypeError(self, a_specification):
+        with pytest.raises(TypeError):
+            a_specification.setTraitProperty("first_trait", "unknown type", object())
+
+
+def test_BlobTrait_traitId():
+    assert trait.BlobTrait.traitId() == "blob"
+
+
+class Test_BlobTrait_isValid:
+    def test_when_wrapping_BlobSpecification_then_returns_true(self, a_blob_specification):
+        assert trait.BlobTrait(a_blob_specification).isValid()
+
+    def test_when_wrapping_blob_supporting_spec_then_returns_true(self):
+        assert trait.BlobTrait(
+            specification.Specification([trait.BlobTrait.traitId(), "other"])).isValid()
+
+    def test_when_wrapping_non_blob_spec_then_returns_false(self, a_specification):
+        assert not trait.BlobTrait(a_specification).isValid()
+
+
+class Test_BlobTrait_getURL:
+    def test_when_spec_doesnt_have_blob_trait_then_raises_IndexError(self, a_specification):
+        with pytest.raises(IndexError):
+            trait.BlobTrait(a_specification).getUrl()
+
+    def test_when_url_in_spec_has_wrong_value_type_then_raises_RuntimeError(self):
+        spec = specification.Specification([trait.BlobTrait.traitId()])
+        spec.setTraitProperty(trait.BlobTrait.traitId(), "url", 123)
+
+        with pytest.raises(RuntimeError):
+            trait.BlobTrait(spec).getUrl()
+
+    def test_when_spec_has_no_url_then_returns_None(self, a_blob_specification):
+        assert trait.BlobTrait(a_blob_specification).getUrl() is None
+
+    def test_when_spec_has_url_then_returns_url(self, a_blob_specification):
+        expected = "some://url"
+        a_blob_specification.setTraitProperty(trait.BlobTrait.traitId(), "url", expected)
+
+        actual = trait.BlobTrait(a_blob_specification).getUrl()
+
+        assert actual == expected
+
+
+class Test_BlobTrait_setURL:
+    def test_when_spec_doesnt_have_blob_trait_then_raises_IndexError(self, a_specification):
+        with pytest.raises(IndexError):
+            trait.BlobTrait(a_specification).setUrl("some://url")
```

Comment:
Does it then make sense to have concrete specification classes, i.e. `BlobSpecification`?  Since you could transform it into a different specification by setting new traits on it.

Reply:
It felt like the concrete specs are more generally a template for consistency. The reason for an `ImageSpecification` being to ensure that all hosts.managers dealing with an image agree it combines `blob`, `image`, `raster` for example, and make it easy for a programmer to know what should/shouldnt be available (as there aren't trait accessors on that spec for traits an "image" doesn't have).

I guess it throws up the Q as to whether:
 - Direct trait use is allowed (feels like it should be) take `register` for example, wanting to just use a `BlobTrait`, and we dont need `BlobSpecification` specifically unless its composes multiple traits.
 - Or, should only `Specifications` be used, and we consequently always have a spec for each trait.

Reply:
I guess, coupled with the other comment above, is this desirable for a manager in resolve:

```python

result = Specification()

if myAsset.filePath:
  blob = BlobTrait(result)
  blob.setUrl(f"file://{myAsset.filePath}")

if myAsset.isSequence:
  range = FrameRangeTrait(result)
  range.setStartFrame(myAsset.startFrame)
...
```

Reply:
.. though it does pose the question of "does just wrapping add the trait ID", maybe we should look at how applied schemas work in USD.

Reply:
From what I read, USD's applied API schemas do indeed add the attributes. Non-applied schemas (which are similar to how I have viewed traits so far) do not.  So, as in all things, USD has it both ways.  I'm not 100% that's how applied vs. non-applied schemas work, but that's my understanding.

Reply:
We can re-visit adding traits later, but definitely need to make sure that we highlight the use of `isValid` in the docs.